src/compile-middleware.ts

```diff
@@ -84,56 +105,207 @@ export function babelCompile(forceCompile: boolean): RequestHandler {
+      } else {
+        transformed = body;
       }
-      babelCompileCache.set(source, body);
-
-      return body;
+      babelCompileCache.set(body, transformed);
+      return transformed;
     },
   });
 }
 
-function compileHtml(source: string, location: string): string {
+function compileHtml(
+    source: string,
+    location: string,
+    componentUrl: string,
+    options: CompileOptions): string {
   const document = parse5.parse(source);
-  const scriptTags = dom5.queryAll(document, isInlineJavaScript);
-  for (const scriptTag of scriptTags) {
-    try {
-      const script = dom5.getTextContent(scriptTag);
-      const compiledScriptResult = compileScript(script);
-      dom5.setTextContent(scriptTag, compiledScriptResult);
-    } catch (e) {
-      // By not setting textContent we keep the original script, which
-      // might work. We may want to fail the request so a better error
-      // shows up in the network panel of dev tools. If this is the main
-      // page we could also render a message in the browser.
-      console.warn(`Error compiling script in ${location}: ${e.message}`);
+  let requireScriptTag, wctScriptTag;
+
+  for (const scriptTag of dom5.queryAll(document, isJsScriptNode)) {
+    // Is this a module script we should transform?
+    const transformingModule = options.transformModules &&
+        dom5.getAttribute(scriptTag, 'type') === 'module';
+
+    if (transformingModule && !requireScriptTag) {
+      // We need RequireJS to load the AMD modules we are declaring. Inject the
+      // dependency as late as possible (right before the first module is
+      // declared) because some of our legacy non-module dependencies,
+      // typically loaded in <head>, behave differently when window.require is
+      // present.
+      const fragment = parse5.parseFragment(
+          `<script src="/${componentUrl}/requirejs/require.js"></script>\n`);
+      requireScriptTag = fragment.childNodes[0];
+      dom5.insertBefore(scriptTag.parentNode, scriptTag, fragment);
+    }
+
+    const src = dom5.getAttribute(scriptTag, 'src');
+    const isInline = !src;
+
+    if (src && src.includes('web-component-tester/browser.js')) {
+      wctScriptTag = scriptTag;
+    }
+
+    if (transformingModule && !isInline) {
+      // Transform an external module script into a `require` for that module,
+      // to be executed immediately.
+      dom5.replace(
+          scriptTag,
+          parse5.parseFragment(`<script>require(["${src}"]);</script>\n`));
+
+    } else if (isInline) {
+      let js = dom5.getTextContent(scriptTag);
+      const plugins = [];
+      if (options.transformES2015) {
+        plugins.push(...es2015Plugins);
+      }
+      if (transformingModule) {
+        plugins.push(...modulesPlugins);
+      }
+
+      let compiled;
+      try {
+        compiled = babelCore.transform(js, {plugins}).code;
+      } catch (e) {
+        // Leave the original script as-is. Maybe it will work...
```

Comment:
Unless the `let compiled` statement above is initialized with the original js value, it will be undefined and this would break on the `.replace` call below or the dom5.setTextContent would receive undefined.